service/frontend/redirection_interceptor.go

```diff
@@ -0,0 +1,254 @@
+	}
+)
+
+type (
+	respAllocFn func() interface{}
+
+	// RedirectionInterceptor is simple wrapper over frontend service, doing redirection based on policy
+	RedirectionInterceptor struct {
+		currentClusterName string
+		config             *Config
+		namespaceCache     namespace.Registry
+		redirectionPolicy  DCRedirectionPolicy
+		logger             log.Logger
+		clientBean         client.Bean
+		metricsHandler     metrics.Handler
+		timeSource         clock.TimeSource
+	}
+)
+
+// NewRedirectionInterceptor creates DC redirection interceptor
+func NewRedirectionInterceptor(
+	configuration *Config,
+	namespaceCache namespace.Registry,
+	policy config.DCRedirectionPolicy,
+	logger log.Logger,
+	clientBean client.Bean,
+	metricsHandler metrics.Handler,
+	timeSource clock.TimeSource,
+	clusterMetadata cluster.Metadata,
+) *RedirectionInterceptor {
+	dcRedirectionPolicy := RedirectionPolicyGenerator(
+		clusterMetadata,
+		configuration,
+		namespaceCache,
+		policy,
+	)
+
+	return &RedirectionInterceptor{
+		currentClusterName: clusterMetadata.GetCurrentClusterName(),
+		config:             configuration,
+		redirectionPolicy:  dcRedirectionPolicy,
+		namespaceCache:     namespaceCache,
+		logger:             logger,
+		clientBean:         clientBean,
+		metricsHandler:     metricsHandler,
+		timeSource:         timeSource,
+	}
+}
+
+var _ grpc.UnaryServerInterceptor = (*RedirectionInterceptor)(nil).Intercept
+
+func (i *RedirectionInterceptor) Intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (_ interface{}, retError error) {
+	defer log.CapturePanic(i.logger, &retError)
+
+	_, methodName := interceptor.SplitMethodName(info.FullMethod)
+	if _, ok := localAPIResults[methodName]; ok {
+		return i.handleLocalAPIInvocation(ctx, req, handler, methodName)
+	}
+	if respAllocFn, ok := globalAPIResults[methodName]; ok {
+		namespaceName := interceptor.GetNamespace(i.namespaceCache, req)
+		return i.handleRedirectAPIInvocation(ctx, req, info, handler, methodName, respAllocFn, namespaceName)
+	}
+
+	// this should not happen, log warn and move on
+	i.logger.Warn(fmt.Sprintf("RedirectionInterceptor encountered unknown API: %v", methodName))
+	return handler(ctx, req)
+}
+
+func (i *RedirectionInterceptor) handleLocalAPIInvocation(
+	ctx context.Context,
+	req interface{},
+	handler grpc.UnaryHandler,
+	methodName string,
+) (_ interface{}, retError error) {
+	scope, startTime := i.beforeCall(dcRedirectionMetricsPrefix + methodName)
+	defer func() {
+		i.afterCall(scope, startTime, i.currentClusterName, retError)
+	}()
+	return handler(ctx, req)
+}
+
+func (i *RedirectionInterceptor) handleRedirectAPIInvocation(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+	methodName string,
+	respAllocFn func() interface{},
+	namespaceName namespace.Name,
+) (_ interface{}, retError error) {
```

Comment:
Do we need/want to avoid multiple redirects? e.g. to prevent a request from bouncing between cells.

Reply:
request will only bounce between clusters if namespace flip flop